src/reverse_offload/mpi_transport.cpp

```diff
@@ -469,568 +357,134 @@ void MPITransport::team_broadcast(void *dst, void *src, int size, int win_id,
-  NET_CHECK(
-      MPI_Group_translate_ranks(grp, grp_size, ranks.data(), world_grp, world_ranks.data()));
-
-  int type_size{};
-  MPI_Datatype mpi_type{convertType(type)};
-  NET_CHECK(MPI_Type_size(mpi_type, &type_size));
-
-  int num_clust = sqrt(pe_size);
-  int clust_size{(pe_size + num_clust - 1) / num_clust};
-  assert(num_clust * clust_size == pe_size);
-  int clust_id{new_rank / clust_size};
-
-  if (MAX_ATA_BUFF_SIZE < type_size * size * pe_size) {
-    fprintf(stderr, "Alltoall size %d exceeds max MAX_ATA_BUFF_SIZE %d\n",
-            type_size * size * pe_size, MAX_ATA_BUFF_SIZE);
-    abort();
-  }
-
-  std::vector<MPI_Request> clust_req(pe_size);
-
-  // Step 1: Send data to PEs in cluster
-  for (int i{0}; i < pe_size; ++i) {
-    int src_offset{(new_rank % clust_size + (i / clust_size) * clust_size) *
-                    type_size * size};
-    int dst_offset{i * type_size * size};
-    NET_CHECK(MPI_Rget(
-        reinterpret_cast<void *>(
-            (reinterpret_cast<char *>(ata_buffptr) + dst_offset)),
-        size, mpi_type, world_ranks[clust_id * clust_size + (i % clust_size)],
-        bp->heap_window_info[win_id]->get_offset(reinterpret_cast<char *>(src) +
-                                                 src_offset),
-        size, mpi_type, bp->heap_window_info[win_id]->get_win(),
-        &clust_req[i]));
+      requests.push_back({req, {nullptr, contextId, false}});
+      outstanding[contextId]++;
   }
 
-  NET_CHECK(MPI_Waitall(pe_size, clust_req.data(), MPI_STATUSES_IGNORE));
-
-  // Step 2: Send final data to PEs outside cluster
-  for (int i{0}; i < num_clust; ++i) {
-    int src_offset{i * type_size * size * clust_size};
-    int dst_offset{clust_id * type_size * size * clust_size};
-    NET_CHECK(MPI_Put(
-        reinterpret_cast<void *>(
-            (reinterpret_cast<char *>(ata_buffptr) + src_offset)),
-        size * clust_size, mpi_type,
-        world_ranks[(new_rank % clust_size) + i * clust_size],
-        bp->heap_window_info[win_id]->get_offset(dst) + dst_offset,
-        size * clust_size, mpi_type, bp->heap_window_info[win_id]->get_win()));
-
-    // Since MPI makes puts as complete as soon as the local buffer is free,
-    // we need a flush to satisfy quiet.
-    NET_CHECK(
-        MPI_Win_flush(world_ranks[(new_rank % clust_size) + i * clust_size],
-                      bp->heap_window_info[win_id]->get_win()));
-  }
-
-  int stride{world_ranks[1] - world_ranks[0]};
-  MPI_Comm comm_cluster{
-      createComm(world_ranks[clust_id * clust_size], stride, clust_size)};
-  MPI_Comm comm_ring{createComm(world_ranks[new_rank % clust_size],
-                                stride * clust_size, num_clust)};
-
-  barrier(contextId, status, false, comm_cluster);
-  barrier(contextId, status, blocking, comm_ring);
+  NET_CHECK(MPI_Win_flush_all(bp->heap_window_info[win_id]->get_win()));
+  barrier(contextId, status, blocking, comm);
+  quiet(contextId,status);
```

Comment:
```suggestion
  barrier(contextId, nullptr, false, comm);
  quiet(contextId, status);
```
to avoid notifying the GPU thread when barrier completes